codefeedr-plugins/codefeedr-travis/src/main/scala/org/codefeedr/plugins/travis/util/TravisBuildCollector.scala

```diff
@@ -0,0 +1,158 @@
+  * @param pollingInterval Interval at which builds are queried
+  * @param timeoutSeconds How long it will wait for the build to be created
+  */
+class TravisBuildCollector(repoOwner: String,
+                           repoName: String,
+                           branchName: String,
+                           pushCommitSha: String,
+                           pushDate: LocalDateTime,
+                           travis: TravisService,
+                           pollingInterval: Int = 30000,
+                           timeoutSeconds: Int = 60) {
+
+  private var minimumStartDate: LocalDateTime = pushDate
+  private var build: Option[TravisBuild] = None
+
+
+  /**
+    * Keeps requesting the build until it is finished
+    * @return A finished Travis build
+    */
+  def requestFinishedBuild(): Future[TravisBuild] = Future {
+    while (!isReady) {
+
+      build = requestBuild
+      checkIfBuildShouldBeKnownAlready()
+
+      if (!isReady) {
+        blocking {
+          Thread.sleep(pollingInterval)
+        }
+      }
+    }
+    build.get
+  }
+
+  def checkIfBuildShouldBeKnownAlready(): Unit = {
+    val waitUntil = pushDate.plusSeconds(100 * timeoutSeconds)
+    if (build.isEmpty && waitUntil.isBefore(LocalDateTime.now(ZoneId.of("GMT")))) {
+      throw BuildNotFoundForTooLongException("Waited " + timeoutSeconds + " seconds for build, but still not found" +
+        ", probably because " + repoOwner + "/" + repoName + "is not active on Travis")
+    }
+  }
+
+  /**
+    * Checks whether or not a build is completed.
+    * @return Whether or not a build is completed
+    */
+  def isReady: Boolean = {
+    if (build.nonEmpty) {
+      val state = build.get.state
+      return state == "passed" || state == "failed" || state == "canceled" || state == "errored"
+    }
+    false
+  }
+
+  /**
+    * Requests the build information based on if a build id is known
+    * @return A Travis build
+    */
+  def requestBuild: Option[TravisBuild] = {
+    try {
+      build match {
+        case None => requestUnknownBuild()
+        case Some(_) => requestKnownBuild()
+      }
+    } catch {
+      case _: CouldNotExtractException =>
+        throw CouldNotAccessTravisBuildInfo("Could not retrieve Travis build info for: "
+          + repoOwner + "/" + repoName)
+      case _: CouldNotGetResourceException =>
+        None
+    }
+  }
+
+  /**
+    * Requests a build from travis of which the build id is known
+    * @return A TravisBuild if it is found, None otherwise
+    */
+  def requestKnownBuild(): Option[TravisBuild] = {
+    assert(build.nonEmpty)
+    Some(travis.getBuild(build.get.id))
+  }
+
+  /**
+    * Looks through the Travis builds of the push event to find the build with the corresponding commit sha.
+    * The builds are sorted on date, so it only looks after the push date
+    * @return A TravisBuild if it is found, None otherwise
+    */
+  def requestUnknownBuild(): Option[TravisBuild] = {
```

Comment:
I'm not sure, but this method looks rather big & complicated. Is it possible to split it in the while loop? Better code hub will probably bitch about it

Reply:
I improved it a little, but I don't know if splitting it up is worth or needed